Spanner/tests/Unit/Connection/GrpcTest.php

```diff
@@ -1418,6 +1419,30 @@ public function larOptions()
         ];
     }
 
+    public function testSerializerCustomEncoder()
+    {
+        $grpc = new GrpcStub();
+        $msg = new MockResponse(['name' => 'foo']);
+        $serializer = new Serializer([], [], [], [], [
+            MockResponse::class => function ($msg) {
+                return ['name' => 'bar'];
+            }
+        ]);
+        $grpc->setSerializer($serializer);
+        $arr = $grpc->callEncodeMessage($msg);
+        $this->assertEquals('bar', $arr['name']);
+    }
+
+    public function testSerializerWithoutCustomEncoder()
+    {
+        $grpc = new GrpcStub();
+        $msg = new MockResponse(['name' => 'foo']);
+        $serializer = new Serializer();
+        $grpc->setSerializer($serializer);
+        $arr = $grpc->callEncodeMessage($msg);
+        $this->assertEquals('foo', $arr['name']);
+    }
+
```

Comment:
These tests don't actually really test anything that we don't already test in GAX. The `GrpcStub` is defining it's own version of the `$serializer` property (since `$serializer` is private in the `Grpc` class), and then calling that version with it's own `callEncodeMessage`. I would suggest removing these tests (and reverting the changes to `GrpcStub`) in favor of a specific test which tests the custom serialization we added. I was able to run this test and it worked:

```suggestion
    public function testPartialResultSetCustomEncoder()
    {
        $partialResultSet = new PartialResultSet();
        $partialResultSet->mergeFromJsonString(json_encode([
            'metadata' => [
                'transaction' => [
                    'id' => base64_encode(0b00010100) // bytedata is represented as a base64-encoded string in JSON
                ],
                'rowType' => [
                    'fields' => [
                        ['type' => ['code' => 'INT64']] // enums are represented as their string equivalents in JSON
                    ]
                ],
            ],
        ]));

        $this->assertEquals(0b00010100, $partialResultSet->getMetadata()->getTransaction()->getId());
        $this->assertEquals(2, $partialResultSet->getMetadata()->getRowType()->getFields()[0]->getType()->getCode());

        // decode the message and ensure it's decoded as expected
        $grpc = new Grpc();
        $serializerProp = new \ReflectionProperty($grpc, 'serializer');
        $serializerProp->setAccessible(true);
        $serializer = $serializerProp->getValue($grpc);
        $arr = $serializer->encodeMessage($partialResultSet);

        // We expect this to be the binary string
        $this->assertEquals(0b00010100, $arr['metadata']['transaction']['id']);
        // We expect this to be the integer
        $this->assertEquals(2, $arr['metadata']['rowType']['fields'][0]['type']['code']);
    }
```

Reply:
Neat.
I never realized we can do this using `mergeFromJsonString`.
Thanks.